### ISSUE TITLE
Manager self-demotion should not be possible

### ISSUE THREAD

**Issue by ehuggett** [iomad/iomad#306]
A Company manager should not be able to re-assign themselves to a lesser role within the company using the "Assign department users/managers" tool as it cannot be undone without contacting the site administrator.

Some checks seem to be in place to prevent this, for example an entry for the company managers own account is not shown when listing the current/potential users for the "staff" or "company manager" role for the company itself (technically, the root department that has the same name as the company).

However, an entry for the managers own user account DOES appear when you select "Department training manager" on the parent company. When you select a sub-department an entry for the managers own account is shown for all three roles ("staff"/"company manager"/"department training manager").

**Comment by ehuggett** [comment]
Ah, correction to the above, its more complicated than I realised.

The above is true when you have just created a new company with 1 Company manager account and one sub-department.

I think this behaviour may hinge on [this](https://github.com/iomad/iomad/blob/IOMAD_29_STABLE/blocks/iomad_company_admin/lib/user_selectors.php#L450)

```
$departmentusers = $this->get_department_user_ids();
        if (count($departmentusers) > 0) {
            $departmentuserids = $USER->id.','.implode(',', $departmentusers);
            if ($departmentuserids != ',') {
                $userfilter = " AND NOT u.id in ($departmentuserids) ";
```

Notice the current user id will only be excluded if there are already users in the department, I think this may explain why an entry for the managers own account is always shown in potential users for an (empty) sub department.

**Comment by ehuggett** [comment]
OK, with one assumption that needs to be double checked I think this fixes the issue (and tidies it up a little).

The assumption is $USER->id will _always_ return the ID of the current user. I am also not sure what effect this will have on other functionality, I haven't check yet!

```
diff --git a/blocks/iomad_company_admin/lib/user_selectors.php b/blocks/iomad_company_admin/lib/user_selectors.php
index 6e42c72..e3c964c 100644
--- a/blocks/iomad_company_admin/lib/user_selectors.php
+++ b/blocks/iomad_company_admin/lib/user_selectors.php
@@ -447,16 +447,10 @@ class potential_department_user_selector extends user_selector_base {
         $countfields = 'SELECT COUNT(1)';

         $departmentusers = $this->get_department_user_ids();
-        if (count($departmentusers) > 0) {
-            $departmentuserids = $USER->id.','.implode(',', $departmentusers);
-            if ($departmentuserids != ',') {
-                $userfilter = " AND NOT u.id in ($departmentuserids) ";
-            } else {
-                $userfilter = "";
-            }
-        } else {
-            $userfilter = "";
-        }
+        // Add the ID of the current User to exclude them from the results
+        $departmentusers[] = $USER->id;
+        $userfilter = " AND NOT u.id in (" . implode(",",$departmentusers) . ") ";
+
         if ($this->roletype != 0) {
             // Dealing with management possibles could be from anywhere.
             $deptids = implode(',', array_keys($this->subdepartments));
```

**Comment by turf212** [comment]
You are right.  $USER->id (if $USER is made global) will return the current users userid.

I'm adding this fix to 3.0

**Comment by turf212** [comment]
Fixed in 3.0
